### ISSUE TITLE
Export individual CTDL-ASN competencies.

### ISSUE THREAD

**Issue by Lomilar** [cassproject/CASS#82]
@stuartasutton: Well, there is no conceptual reason that an export a the level of a single competency should not be possible (hopefully with the addition of the context statement). Originally, download was only for the individual URI requested (even the framework URI). I.e., there was no way to get it all. That was changed so that a request to download the framework yielded the full framework + competencies graph.  I think considerable water has gone under the bridge with context etc. since then. In terms of usefulness, retrieving by individual node or by the full graph should be supported if we can make sense of it in terms of user experience.

**Comment by stuartasutton** [comment]
Fritz, this has nothing to do with sharing competencies between frameworks and the 2 and 3 entity model.It's merely asking whether the competency node can be exported individually.

**Comment by Lomilar** [comment]
If it were exported without the framework as context, it would be exported without connections to other competencies (no isChild/hasChild/other relations)

Still, it can be supported.

**Comment by stuartasutton** [comment]
Fritz, I'm not getting this. If I resolve the URI http://asn.desire2learn.com/resources/S2695991, I should get back this:

```
<http://asn.desire2learn.com/resources/S2695991> rdf:type asn:Statement ;
     dcterms:isPartOf <http://asn.desire2learn.com/resources/D2695955> ;
      asn:authorityStatus <http://purl.org/ASN/scheme/ASNAuthorityStatus/Original> ;
      asn:indexingStatus <http://purl.org/ASN/scheme/ASNIndexingStatus/No> ;
      dcterms:educationLevel <http://purl.org/ASN/scheme/ASNEducationLevel/AdultEducation> ,                                                              
           <http://purl.org/ASN/scheme/ASNEducationLevel/VocationalTraining> ,                                                                
           <http://purl.org/ASN/scheme/ASNEducationLevel/ProfessionalEducation-Development> ;
      dcterms:subject <http://purl.org/ASN/scheme/ASNTopic/careerEducation> ,
            <http://purl.org/ASN/scheme/ASNTopic/technology> ;
     asn:statementLabel "Topic Cluster"@en-US ;
     dcterms:description "Fundamentals of Resource Description Framework"@en-US ;
     dcterms:language <http://id.loc.gov/vocabulary/iso639-2/eng> ;
     gemq:hasChild <http://asn.desire2learn.com/resources/S2696001> ,
           <http://asn.desire2learn.com/resources/S2696007> ,
           <http://asn.desire2learn.com/resources/S2696012> ,
           <http://asn.desire2learn.com/resources/S2696017> ;
     gemq:isChildOf <http://asn.desire2learn.com/resources/D2695955> .
```

**Comment by Lomilar** [comment]
Correct, in a three object model, the linking statements are separate from the descriptive statements.

```
<http://asn.desire2learn.com/resources/S2695991> rdf:type asn:Statement ;
      asn:authorityStatus <http://purl.org/ASN/scheme/ASNAuthorityStatus/Original> ;
      asn:indexingStatus <http://purl.org/ASN/scheme/ASNIndexingStatus/No> ;
      dcterms:educationLevel <http://purl.org/ASN/scheme/ASNEducationLevel/AdultEducation> ,                                                              
           <http://purl.org/ASN/scheme/ASNEducationLevel/VocationalTraining> ,                                                                
           <http://purl.org/ASN/scheme/ASNEducationLevel/ProfessionalEducation-Development> ;
      dcterms:subject <http://purl.org/ASN/scheme/ASNTopic/careerEducation> ,
            <http://purl.org/ASN/scheme/ASNTopic/technology> ;
     asn:statementLabel "Topic Cluster"@en-US ;
     dcterms:description "Fundamentals of Resource Description Framework"@en-US ;
     dcterms:language <http://id.loc.gov/vocabulary/iso639-2/eng> ;
```
is separate from
```
<http://asn.desire2learn.com/resources/S2695991> rdf:type asn:Statement ;
     dcterms:isPartOf <http://asn.desire2learn.com/resources/D2695955> ;
     gemq:hasChild <http://asn.desire2learn.com/resources/S2696001> ,
           <http://asn.desire2learn.com/resources/S2696007> ,
           <http://asn.desire2learn.com/resources/S2696012> ,
           <http://asn.desire2learn.com/resources/S2696017> ;
     gemq:isChildOf <http://asn.desire2learn.com/resources/D2695955> .
```
making the first document reusable in different versions of the framework or different frameworks without changing the IDs.

The same thing could be accomplished by duplicating the statement with a different ID and using http://purl.org/ASN/schema/core/exactAlignment to link them when a different version is published or a derivative framework is created.

The solution we're proposing will identify in the URL the description document (the first one) and the linking document found in the framework (the second one) to export a complete ctdlasn competency.

**Comment by siuc-nate** [comment]
If you're reusing the same competency data with multiple frameworks (which I'm not sure we want/need to do to begin with), then what about just retrieving _all_ of the connections for that competency from all frameworks it appears in?  

That would allow you to populate the hasChild/isChild properties with data from 1-n frameworks' linking objects (in the vast majority of cases this will just be one framework), and I _think_ still be semantically correct if that data comes from more than one framework.

In other words, say you have 2 frameworks and some number of competencies where one competency appears in both frameworks. Retrieving the data for any of those competencies would also include the hasChild/isChildOf data from all frameworks that that competency appears in. The only time more than one framework's data would appear would be if you retrieved the data for the competency that appears in multiple frameworks.

**Comment by stuartasutton** [comment]
Is this merely a philosophical discussion or does it have project significant consequences?  I thought I had put a stake in the ground some time ago in the discussions about change in competency frameworks that, at least in the CER, the a competency cannot appear in more than one framework unless it is in the exact same taxon path--the same hierarchy from framework to leaf. So, even if you are reusing a free-floating competency it could be the basis for use in multiple disparate hierarchies, BUT, none of those derivatives should have the same URI as the free-floater because they are all contextually different. Obviously, a relationship should be declared between the derivatives and the free-floater, but shared URI? I don't think so.

**Comment by Lomilar** [comment]
>if a competency, even as a free-floating entity, only ever appears in one framework (because we duplicate it when it appears in another framework and assign unique URIs to each) then I think (correct me if I'm wrong) it would have the same end result (in terms of the CTDL-ASN export) as attaching the hasChild/isChildOf data to the competency itself internally.

^ Correct. There isn't a functional difference because the CTDL-ASN profile of the editor doesn't ([or shouldn't](https://github.com/cassproject/cass-editor/issues/370)) permit competency nodes to be shared between frameworks, just copied.

Honestly, I was just trying to explain why we hadn't already allowed this functionality, and it snowballed a little bit.

**Comment by stuartasutton** [comment]
It has indeed drifted (snowballed). The original note from me the very top had nothing to do with ANYTHING other than being able to retrieve (resolve) a specific competency by URI in the editor.

**Comment by siuc-nate** [comment]
Getting back to:

>We can support such an export, but we need to know which framework the competency is being taken in context of.
>Since this is a feature of the export URL and not the `@id` of the exported object, this is possible within that context.

My understanding is that it basically _should be_ a feature of the `@id` of the exported object (in the sense that no two frameworks can share a competency with the same `@id`).

A quick example based on my suggestion above (and because i had it nearly typed out anyway as the two posts above this one appeared):
```
//Frameworks
{ "@id": "F1" }
{ "@id": "F2" }

//Competencies
{ "@id": "C1" }
{ "@id": "C2" }
{ "@id": "C3" }
{ "@id": "C4" }
{ "@id": "C5" }

//Linking Objects
{ "@id": "L1", "link": "C1-isTopChildOf->F1" }
{ "@id": "L3", "link": "C4-isTopChildOf->F2" }
{ "@id": "L4", "link": "C2-isChildOf->C1" }
{ "@id": "L5", "link": "C3-isChildOf->C2" }
{ "@id": "L6", "link": "C4-isChildOf->C2" }
{ "@id": "L7", "link": "C5-isChildOf->C4" }
```
When you export a competency in CTDL-ASN, query the links for anything that relates to that competency and assemble the JSON appropriately before returning it.

Resolve `C2` and get:
```
{ 
  "@id": "C2",
  "hasChild: [ "C3", "C4" ],
  "isChildOf": [ "C1" ], 
  "isPartOf": [ "F1" ]
}
```

Resolve `C4` and get:
```
{ 
  "@id": "C4", 
  "isChildOf": [ "C2" ], 
  "isPartOf": [ "F1", "F2" ], 
  "isTopChildOf": [ "F2" ]
}
```

Note that because of C4 being allowed to show up in two frameworks (via the `because of the C4-isChildOf-C2` link:
- C4 appears as a child of C2 even though it isn't _directly_ (via linking objects) listed as a member of F1
- C4 appears as a part of both F1 and F2

You can avoid that mess by never letting a competency appear in more than one framework - always duplicate it (again, as mentioned above).  Then you'd have:
```
//Frameworks
{ "@id": "F1" }
{ "@id": "F2" }

//Competencies
{ "@id": "C1" }
{ "@id": "C2" }
{ "@id": "C3" }
{ "@id": "C4a" }
{ "@id": "C4b" }
{ "@id": "C5" }

//Linking Objects
{ "@id": "L1", "link": "C1-isTopChildOf->F1" }
{ "@id": "L3", "link": "C4a-isTopChildOf->F2" }
{ "@id": "L4", "link": "C2-isChildOf->C1" }
{ "@id": "L5", "link": "C3a-isChildOf->C2" }
{ "@id": "L6", "link": "C4b-isChildOf->C2" }
{ "@id": "L7", "link": "C5-isChildOf->C4a" }
```

Resolve `C2` and get:
```
{ 
  "@id": "C2",
  "hasChild: [ "C3", "C4b" ],
  "isChildOf": [ "C1" ], 
  "isPartOf": [ "F1" ]
}
```

Resolve `C4a` and get:
```
{ 
  "@id": "C4a", 
  "hasChild": [ "C5" ],
  "isPartOf": [ "F2" ], 
  "isTopChildOf": [ "F2" ]
}
```

That essentially means that if you have an `@id` of "C4a", you don't _need_ the `@id` of "F2" because you can figure that out by querying your linking objects - meaning you shouldn't need a URI like `/ctdlasn/<frameworkCTID>/<competencyCTID>` (though maybe that is faster and/or allows the framework's `@id` to act as a filter so no other frameworks' `@id`s are allowed to show up?)

Anyway, as I said, I had this mostly typed out when the previous two posts appeared so maybe we're already on the same page.  I figure I'll put this out there anyway just to double-check.

**Comment by Lomilar** [comment]
Done on server side. cassproject/cass-editor#449 now responsible for this feature.
